language/move-vm/runtime/src/loader.rs

```diff
@@ -217,39 +238,45 @@ impl ModuleCache {
         }
     }
 
-    fn add_module(&mut self, natives: &NativeFunctions, module: &CompiledModule) -> VMResult<()> {
+    fn add_module(
+        &mut self,
+        natives: &NativeFunctions,
+        link_context: AccountAddress,
+        module: &CompiledModule,
+    ) -> VMResult<()> {
         let starting_idx = self.structs.len();
         for (idx, struct_def) in module.struct_defs().iter().enumerate() {
             let st = self.make_struct_type(module, struct_def, StructDefinitionIndex(idx as u16));
             self.structs.push(Arc::new(st));
         }
-        self.load_field_types(module, starting_idx).map_err(|err| {
-            // clean up the structs that were cached
-            self.structs.truncate(starting_idx);
-            err.finish(Location::Undefined)
-        })?;
+        self.load_field_types(link_context, module, starting_idx)
```

Comment:
to continue with my drama (who knows if you will ever read those comments in the order they were made.. lol), this is still under lock and `load_field_types` can end up walking through all the types in the module to find the field type. So a possible O(fields_count * types_count) complexity, or quadratic to the limits on both.
This is not something that this PR made (it is courtesy of yours truly) but something that may be amplified in a chain of dependencies loaded under lock

Reply:
Yeah, I noticed this... bring on the arena!